copier/main.py

```diff
@@ -388,7 +388,14 @@ def _execute_tasks(self, tasks: Sequence[Task]) -> None:
 
             extra_env = {k[1:].upper(): str(v) for k, v in extra_context.items()}
             with local.cwd(working_directory), local.env(**extra_env):
-                subprocess.run(task_cmd, shell=use_shell, check=True, env=local.env)
+                try:
+                    subprocess.run(task_cmd, shell=use_shell, check=True, env=local.env)
+                except subprocess.CalledProcessError as error:
+                    if task.failure_message:
+                        message = self._render_string(task.failure_message, extra_context)
+                    else:
+                        message = str(error)
+                    raise UserMessageError(f"Task {i + 1} failed: {message}")
```

Comment:
This may be a breaking change for developers who integrate Copier via Python API and expect a `subprocess.CalledProcessError` to be raised upon task failure. I think we can make this backwards compatible once #1991 has been merged by extending the new `TaskError` exception by an optional constructor argument `message: str | None = None` and passing the failure message. Then, CLI users see this message (if provided) and API users can catch `subprocess.CalledProcessError` as before (and ideally migrate to catching `TaskError` instead).